### PR TITLE
[WIP] use UnstructuredObject type for backup/restore

### PR DESCRIPTION
**Pull request by skriss**
Signed-off-by: Steve Kriss <steve@heptio.com>

This PR addresses https://github.com/heptio/ark/pull/834#issuecomment-420047442.  It adds a new `UnstructuredObject` interface that embeds a `runtime.Unstructured` as well as a `metav1.Object`, both of which `*unstructured.Unstructured` implement.  The benefit of this is that standard Kube metadata is now directly accessible from the object, and we don't have to get a new `meta.Accessor` to get at it easily.  I'd like to do this in v0.10 since we're making lots of plugin changes anyway.

Question: what package do you all think the `UnstructuredObject` interface should go in? I threw it in `pkg/util/kube` for now but I don't think that's where it should live.

Still need to manually test.

### CODE REVIEW

**Comment by nrb** [comment]
> Question: what package do you all think the UnstructuredObject interface should go in? I threw it in pkg/util/kube for now but I don't think that's where it should live.

I don't think I see a better place in the current package layout, so it would likely be a new package, in my mind.

**Comment by skriss** [comment]
Eh, maybe I'll just move `pkg/util/kube` to `pkg/kube` and keep it in there. I haven't had any other great ideas.

**Comment by nrb** [comment]
> Eh, maybe I'll just move pkg/util/kube to pkg/kube and keep it in there. I haven't had any other great ideas.

That seems reasonable to me. In general, I think moving things out of a `util` dir or package is a good idea.

**Comment by carlisia** [comment]
I too vote for moving things out of `util` whenever possible! In this case, would it make sense to name the package `runtime` (under `pkg` and name the object just `Unstructured`? This would reflect what it actually is, `runtime.Unstructured`.

Reviewing the rest...

File: pkg/backup/item_backupper.go [hunk 1]
```diff
@@ -124,14 +124,9 @@ type defaultItemBackupper struct {
...
 // namespaces IncludesExcludes list.
-func (ib *defaultItemBackupper) backupItem(logger logrus.FieldLogger, obj runtime.Unstructured, groupResource schema.GroupResource) error {
-	metadata, err := meta.Accessor(obj)
-	if err != nil {
-		return err
-	}
-
-	namespace := metadata.GetNamespace()
-	name := metadata.GetName()
+func (ib *defaultItemBackupper) backupItem(logger logrus.FieldLogger, obj kube.UnstructuredObject, groupResource schema.GroupResource) error {
```
[the remark this answers is not in the document]
Comment (nrb) [reply pkg/backup/item_backupper.go]: I don't think those names are quite accurate. It's the kubernetes object (like Pod, Event, Secret, ConfigMap, etc) we're trying to backup.

[the remark this answers is not in the document]
Comment (carlisia) [reply pkg/backup/item_backupper.go]: Ah, I see that now.

How about `resource`? I wouldn't mind `obj` being that it is actually, like you say, an actual object. But it is such a loaded name.

File: pkg/backup/backup_test.go [hunk 2]
```diff
@@ -68,12 +67,8 @@ func newFakeAction(resource string) *fakeAction {
 	return (&fakeAction{}).ForResource(resource)
 }
 
-func (a *fakeAction) Execute(item runtime.Unstructured, backup *v1.Backup) (runtime.Unstructured, []ResourceIdentifier, error) {
-	metadata, err := meta.Accessor(item)
-	if err != nil {
-		return item, a.additionalItems, err
-	}
-	a.ids = append(a.ids, kubeutil.NamespaceAndName(metadata))
+func (a *fakeAction) Execute(item kubeutil.UnstructuredObject, backup *v1.Backup) (kubeutil.UnstructuredObject, []ResourceIdentifier, error) {
```
[the remark this answers is not in the document]
Comment (nrb) [reply pkg/backup/backup_test.go]: I agree - we should be consistent with the alias.

File: pkg/backup/item_backupper.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nrb) [reply pkg/backup/item_backupper.go]: `resource` seems fine to me. It's also sometimes shortened to `res` in upstream code. I've seen `res` and `obj` often.

[the remark this answers is not in the document]
Comment (carlisia) [reply pkg/backup/item_backupper.go]: When I see `res` I think it's short for `result` :)

[the remark this answers is not in the document]
Comment (nrb) [reply pkg/backup/item_backupper.go]: That's fair. Honestly I don't think either `resource` or `object` is super descriptive, but they *are* generic resources/objects that are coming through.

**Comment by skriss** [comment]
FYI I'm probably going to hold off on this for now, as I'd like to put a little more thought into getting this right across the entire codebase (including the types our dynamic clients return/accept).

**Comment by nrb** [comment]
> FYI I'm probably going to hold off on this for now,

Would a label help on this? So we can see more clearly on the GH issue list? Or maybe moving back on the ZH lanes?

**Comment by skriss** [comment]
@ncdc I started down this path of defining a new interface for the unstructured objects that we get back from the dynamic client:
```go
// UnstructuredObject is a Kubernetes object stored as a
// map[string]interface{}, with object metadata accessors.
type UnstructuredObject interface {
	runtime.Unstructured
	metav1.Object
}
```
Since we use a lot of the methods that `metav1.Object` defines.

But, I'm wondering if we should just use `*unstructured.Unstructured` everywhere instead.  Any thoughts on this?  Ultimately, I would like whatever we choose to also be the type we use in the plugin interfaces for item actions as well.

**Comment by ncdc** [comment]
@skriss maybe we should start by enumerating what functionality we need from the object type (be it an interface or a struct). Would that help?

**Comment by skriss** [comment]
The motivation for the change was specifically so that we don't have to construct a `meta.Accessor` every time we want to use the methods defined on `*unstructured.Unstructured` / `metav1.Object` to access metadata about an object (`GetNamespace()`, `GetName()`, `GetLabels()`, etc).  These helpers are "hidden" to our code since we're passing `runtime.Unstructured`, even though the concrete type is always an `*unstructured.Unstructured`.  Does that clarify?

**Comment by ncdc** [comment]
Yes, thanks. If we switch to `*unstructured.Unstructured`, the only downside I can think of is that we're at the mercy of apimachinery... If the struct's fields/methods are changed in newer versions, we have to change all of Ark's code, which presumably would trickle down to plugin authors too.

If we do an interface, it could help future-proof and avoid breaking API changes...

WDYT?

**Comment by skriss** [comment]
Yeah, that makes sense and was kind of why I started by defining this interface.  I ended up questioning whether there was actually value in doing so, but I think it makes sense.  Though, now I'm wondering if we should just define it explicitly with exactly the methods we need rather than embedded upstream interfaces, e.g.:
```go
type UnstructuredObject interface {
    UnstructuredContent() map[string]interface{}
    GetNamespace() string
    GetName() string
    ...
}
```
Thoughts on this?

**Comment by skriss** [comment]
Closing this out, may revisit in a new PR.

**Pull request closed without merging**